papermill/engines.py

```diff
@@ -0,0 +1,224 @@
+import copy
+import datetime
+import dateutil
+
+from functools import wraps
+
+# tqdm creates 2 globals lock which raise OSException if the execution
+# environment does not have shared memory for processes, e.g. AWS Lambda
+try:
+    from tqdm import tqdm
+
+    no_tqdm = False
+except OSError:
+    no_tqdm = True
+
+from .preprocess import PapermillExecutePreprocessor
+from .iorw import write_ipynb
+
+
+def catch_nb_assignment(func):
```

Comment:
could you add a comment about when/why this is needed? I think I understand what the code does but I am not sure I grok why it is needed or when it is needed.

Reply:
Definitely. I'll do a comments and tests pass next